### PR TITLE
run all PreFilter when the preemption will happen later in the same scheduling cycle

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind bug
/triage accepted
/priority important-soon

#### What this PR does / why we need it:

Not all PreFilter plugins aren't executed either 
- when one PreFilter plugin returns Unschedulable 
- when PreFilter plugin(s) filter out all nodes by PreFilterResult

But maybe their Filter()s are executed in the preemption and could cause error by trying to read data, which is supposed to be stored by PreFilter (actually not), from cycle state.

---

The former case shouldn't happen in the default scheduler since we don't have PreFilter plugins returning Unschedulable.
But, the latter case may happen in the default scheduler, and cherry-pick is needed.

---

https://github.com/kubernetes/kubernetes/pull/119777 adds integ test for this scenario and https://github.com/kubernetes/kubernetes/pull/119780 proves this patch fixing the bug.

#### Which issue(s) this PR fixes:

Part of #119770

#### Special notes for your reviewer:

PreFilterResult was introduced at 1.24. So, we need to cherry-pick this PR for all supported versions.

#### Does this PR introduce a user-facing change?

```release-note
When PreFilterResult filters out some Nodes, the scheduling framework assumes them as rejected via `UnschedulableAndUnresolvable`, 
that is those nodes won't be in the candidates of preemption process.
Also, corrected how the scheduling framework handle Unschedulable status from PreFilter. 
Before this PR, if PreFilter return `Unschedulable`, it may result in an unexpected abortion in the preemption, 
which shouldn't happen in the default scheduler, but may happen in schedulers with a custom plugin.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sanposhiho** [comment]
will soon add UTs

File: pkg/scheduler/framework/runtime/framework.go
Comment (Huang-Wei) [inline pkg/scheduler/framework/runtime/framework.go:659]: Comparing to always continue running PreFilter upon `Unschedulable`, I'm thinking of _conditional_ continue/stop: 

- option 1: offer a knob in scheduler config to indicate whether it's intended to continue run PreFilter plugin upon the first `Unschedulable` - it should be a profile-specific parameter.
- option 2: try to deduce the user's intent w/o offering a config knob. Technically, it's doable to tell whether a PreFilter plugin returns nil in PreFilterExtensions. For example, suppose pluginA and pluginB both implement PreFilter, pluginB is placed after pluginA. When pluginA fails with `Unschedulable`
  - runPreFilterPlugins() stops if pluginB returns nil PreFilterExtensions
  - runPreFilterPlugins() continues if pluginB returns non-nil PreFilterExtensions

Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go:659]: sg. I prefer option2, as option1 seems a bit difficult for users to understand appropriately, and I don't come up with any scenario that option2 cannot cover, but people want to continue or stop running other PreFilter.

---

But, can we leave this enhancement to a follow-up? because I'm thinking this PR deserves to be cherry-picked to older versions.

Comment (Huang-Wei) [reply pkg/scheduler/framework/runtime/framework.go:659]: cc @alculquicondor @ahg-g @kerthcet A simple approach is to keep running all PreFilter plugins upon Unschedulable, which should be no penalty for in-tree plugins (at least for now).

Comment (Huang-Wei) [reply pkg/scheduler/framework/runtime/framework.go:659]: btw: even for out-of-tree plugins, it's still very strict to trigger this issue, basically you have to satisfy the following conditions:

https://github.com/kubernetes/kubernetes/issues/119770#issuecomment-1666664335

> - an out-of-tree pluginA returns Unschedulable in PreFilter
> - an out-of-tree pluginB is placed after pluginA, and
>   - it implements PreFilter, and write its own key to cycleState
>   - it implements PreFilterExtensions#Add/RemovePod, and report error upon missing key

Comment (sanposhiho) [reply pkg/scheduler/framework/runtime/framework.go:659]: Sorry, I haven't got time to check the sig-meeting rec, but is the conclusion just to go with keeping running all PreFilter when Unschedulable and leave the idea of conditional continue at least for now?
Also, is there any blocker on this PR that I need to pile up to get `/lgtm`?

Comment (Huang-Wei) [reply pkg/scheduler/framework/runtime/framework.go:659]: I'm fine with the current approach. And it's not that severe comparing to other 2 regressions we recently discussed. Let's polish this (I guess there will be conflicts on tests) after those 2 regression PRs get merged first, and then discuss if we want to cherry-pick.

**Comment by sanposhiho** [comment]
Rebased.

**Comment by alculquicondor** [comment]
> when PreFilter plugin(s) filter out all nodes by PreFilterResult

In which case can this happen in the default scheduler?

**Comment by sanposhiho** [comment]
@alculquicondor 
> In which case can this happen in the default scheduler?

We have two PreFilter plugins using PreFilterResult.
This scenario could happen, e.g.,
1. There are 3 nodes.
2. NodeAffinity filters out Node1 and Node2 via PreFilterResult.
3. VolumeBinding filters out Node3 via PreFilterResult.

File: pkg/scheduler/framework/runtime/framework.go
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:676]: There is little value in calling this function, because every branch inside here looks different. Just compare the codes directly.

File: pkg/scheduler/framework/runtime/framework.go
Comment (alculquicondor) [inline pkg/scheduler/framework/runtime/framework.go:702]: I think we should consider the Nodes that aren't returned in the PreFilterResult as UnschedulableAndUnresolvable (I think that applies to the existing plugins that implement it). Or let the plugins give a result.

**Review by alculquicondor — COMMENTED** [review]
General questions:

- What is in the diagnosis when PreFilter returns a subset of nodes?
https://github.com/kubernetes/kubernetes/blob/55f2bc10435160619d1ece8de49a1c0f8fcdf276/pkg/scheduler/schedule_one.go#L174
- Is this accurate? https://github.com/kubernetes/kubernetes/blob/55f2bc10435160619d1ece8de49a1c0f8fcdf276/pkg/scheduler/schedule_one.go#L456-L460

File: pkg/scheduler/framework/runtime/framework.go
Comment (Huang-Wei) [reply pkg/scheduler/framework/runtime/framework.go:702]: Good point. This would benefit the preemption performance of the combination of a plugin that returns non-nil PreFilterResult (e.g., ResourceFit for deamonset pod) + a plugin that return Unschedulable. So we can narrow down the search scope of preemption candidates to the union of PreFilterResult(s); instead of going through all nodes for preemption.

**Comment by sanposhiho** [comment]
> What is in the diagnosis when PreFilter returns a subset of nodes?

If you mean a scenario like:
1. Three Nodes in the cluster. 
2. Node1 and Node2 are filtered out by PreFilterResult.
3. Node3 is filtered out by Filter plugin.

Currently, the diagnosis doesn't get anything for Node1 and Node2. It means, Node1 and Node2 will be the candidates of the preemption. (See [here](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/preemption/preemption.go#L419-L426))

So, after addressing [this](https://github.com/kubernetes/kubernetes/pull/119779#discussion_r1415691836), we have to do this, which is a new change at the latest commit in this PR.
https://github.com/kubernetes/kubernetes/blob/9109e8245a67b9d471b98d832a9ae10a429f7c73/pkg/scheduler/schedule_one.go#L470-L478

---

> Is this accurate?

Looks accurate to me though, is there any part you're doubting?

File: pkg/scheduler/scheduler.go
Comment (sanposhiho) [inline pkg/scheduler/scheduler.go:143]: EvaluatedNodes is only used in logging, not being used at anywhere else like any functionalities or metrics. So, this change doesn't cause anything bad.
https://github.com/kubernetes/kubernetes/blob/6cf77098aaf250697abc1d97fc9e576903b95868/pkg/scheduler/schedule_one.go#L304
(It leads us to another question; whether this `EvaluatedNodes` is worthwhile to keep or not in the first place though)

Comment (alculquicondor) [reply pkg/scheduler/scheduler.go:143]: I've find this information useful when debugging in the past.

**Review by alculquicondor — APPROVED** [review]

**Comment by alculquicondor** [comment]
Please update the release note to reflect that this is a new feature that external plugins can take advantage of. It's not a bug for the existing in-tree plugins.

/lgtm
/hold

**Comment by sanposhiho** [comment]
@alculquicondor Updated. Also I remarked on a bug that could happen in custom schedulers though, eliminate that part if you don't prefer to include it.

**Comment by alculquicondor** [comment]
We are going to need an updated release note. And then prepare the cherry-pick, please.

File: pkg/scheduler/schedule_one_test.go
Comment (alculquicondor) [inline pkg/scheduler/schedule_one_test.go:2230]: Maybe this number shouldn't have changed?

Although it's only useful for debugging.

Comment (sanposhiho) [reply pkg/scheduler/schedule_one_test.go:2230]: It should be changed.

Please refer to the change in pkg/scheduler/scheduler.go in this PR: After this PR, EvaluatedNodes contains the number of nodes that filtered out by PreFilterResult.
https://github.com/kubernetes/kubernetes/pull/119779/files#r1439161728

Comment (alculquicondor) [reply pkg/scheduler/schedule_one_test.go:2230]: Yes, I understand that it's not a bug. However, it was useful to have the value 1 to know that only one node passed PreFilter.

**Pull request merged**
